core/node/node_sync/src/batch_transaction_updater/l1_transaction_verifier.rs

```diff
@@ -0,0 +1,390 @@
+};
+
+/// Verifies the L1 transaction against the database and the SL.
+#[derive(Debug)]
+pub struct L1TransactionVerifier {
+    diamond_proxy_addr: Address,
+    /// ABI of the ZKsync contract
+    contract: ethabi::Contract,
+    pool: ConnectionPool<Core>,
+    pub validate_logs_from_protocol_version: ProtocolVersionId,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum TransactionValidationError {
+    #[error("Batch {batch_number} is not found in the database")]
+    BatchNotFound { batch_number: L1BatchNumber },
+    #[error("Batch transaction {tx_hash} failed on SL")]
+    TransactionFailed { tx_hash: H256 },
+    #[error("Database error")]
+    DatabaseError(#[from] zksync_dal::DalError),
+    #[error("SL client error")]
+    SlClientError(#[from] zksync_eth_client::EnrichedClientError),
+    #[error("Batch tranasction invalid: {reason}")]
+    BatchTransactionInvalid { reason: String },
+    #[error("Other validaion error")]
+    OtherValidationError(#[from] anyhow::Error),
+}
+
+pub fn get_param(log_params: &[ethabi::LogParam], name: &str) -> Option<ethabi::Token> {
+    log_params
+        .iter()
+        .find(|param| param.name == name)
+        .map(|param| param.value.clone())
+}
+
+impl L1TransactionVerifier {
+    pub fn new(diamond_proxy_addr: Address, pool: ConnectionPool<Core>) -> Self {
+        Self {
+            diamond_proxy_addr,
+            contract: zksync_contracts::hyperchain_contract(),
+            pool,
+            validate_logs_from_protocol_version: ProtocolVersionId::Version29,
+        }
+    }
+
+    async fn get_db_batch_metadata(
+        &self,
+        batch_number: L1BatchNumber,
+    ) -> Result<L1BatchMetadata, TransactionValidationError> {
+        match self
+            .pool
+            .connection_tagged("sync_layer")
+            .await?
+            .blocks_dal()
+            .get_l1_batch_metadata_only(batch_number)
+            .await?
+        {
+            Some(batch) => Ok(batch),
+            None => {
+                tracing::debug!(
+                    "Batch {} is not found in the database. Cannot verify transaction right now",
+                    batch_number
+                );
+                Err(TransactionValidationError::BatchNotFound { batch_number })
+            }
+        }
+    }
+
+    async fn should_perform_logs_validation(
+        &self,
+        batch_number: L1BatchNumber,
+    ) -> Result<bool, TransactionValidationError> {
+        match self
+            .pool
+            .connection_tagged("sync_layer")
+            .await?
+            .blocks_dal()
+            .get_batch_protocol_version_id(batch_number)
+            .await?
+        {
+            Some(version) => Ok(version >= self.validate_logs_from_protocol_version),
+            None => {
+                tracing::debug!(
+                    "Batch {} is not found in the database. Cannot verify transaction right now",
+                    batch_number
+                );
```

Comment:
Can this happen? If so, it may make sense to categorize errors as transient and persistent (this one is transient), so `TransactionValidationError` would actually make sense (perhaps with less variants; e.g., all internal errors can be flattened to a single variant: `Internal(#[from] anyhow::Error)`, and persistent validation errors may have a single variant as well: `Validation(anyhow::Error)`).